### PR TITLE
Refactor --allow-ips to handle custom ip-ranges

### PR DESCRIPTION
**Pull request by sjeohp**
Filter some additional reserved ranges from the predefined `public` and `private` lists, as in [#5872](https://github.com/paritytech/parity/issues/5872).  Also:

`--allow-ips FILTER` supports everything it did before but accepts more args:
> FILTER can be one of:
                                   private - connect to private network IP addresses only;
                                   public - connect to public network IP addresses only;
                                   all - connect to any IP address;
                                   none - block all (for use with a custom filter as below);
                                   a custom filter list in the format: "private,ip_range1,-ip_range2,...".
                                   Where ip_range1 would be allowed and ip_range2 blocked;
                                   Custom blocks ("-ip_range") override custom allows ("ip_range");
                                   (default: {flag_allow_ips}).

Accepts one predefined symbol (public, private, all, or none) and unlimited custom ranges.  Could potentially support combinations of predefined like "public private ..." but not sure it would be useful/worthwhile.

### CODE REVIEW

**Comment by arkpar** [comment]
Please re-format changed files to use tabs instead of spaces. We use tabs consistently in our codebase.

File: util/network/src/lib.rs [hunk 1]
```diff
@@ -145,14 +148,61 @@ impl NonReservedPeerMode {
...
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "ipc", binary)]
+pub struct IpFilter {
+    pub predefined: AllowIP,
+    pub custom_allow: Vec<IpNetwork>,
+    pub custom_block: Vec<IpNetwork>,
+} 
+
+impl IpFilter {
+    pub fn new() -> IpFilter {
```
Comment (arkpar) [inline util/network/src/lib.rs]: I'd name it `default`  or `none` or `empty` to show that it creates an empty filter

File: util/network/src/lib.rs [hunk 2]
```diff
@@ -145,14 +148,61 @@ impl NonReservedPeerMode {
...
+        let mut filter = IpFilter::new();
+        for f in s.split_whitespace() {
+            match f {
+                "all" => filter.predefined = AllowIP::All,
+                "private" => filter.predefined = AllowIP::Private,
+                "public" => filter.predefined = AllowIP::Public,
+                "none" => filter.predefined = AllowIP::Non,
+                custom => {
+                    if custom.starts_with("-") {
+                        match IpNetwork::from_str(&custom.to_owned().split_off(1)) {
```
Comment (arkpar) [inline util/network/src/lib.rs]: Could be just `filter.custom_block.push(IpNetwork::from_str(&custom.to_owned().split_off(1)?)`; Same for the `match` below.

File: util/network/src/lib.rs [hunk 3]
```diff
@@ -145,14 +148,61 @@ impl NonReservedPeerMode {
...
+#[derive(Clone, Debug, PartialEq, Eq)]
 pub enum AllowIP {
 	/// Connect to any address
 	All,
 	/// Connect to private network only
 	Private,
 	/// Connect to public network only
 	Public,
+
+    Non,
```
Comment (arkpar) [inline util/network/src/lib.rs]: `None`?

**Comment by arkpar** [comment]
Looks great apart from minor style issues.

File: util/network/src/lib.rs (hunk 3 above)
Comment (sjeohp) [inline util/network/src/lib.rs]: No, I didn't want it to be confused with `Option::None` but reading it now I can see that it might be more confusing :-p

**Pull request merged**